pkg/diff/controller/controller.go

```diff
@@ -191,18 +227,84 @@ func runWorker(ctx context.Context, logger logrus.FieldLogger, taskPool *channel
 				return
 			}
 
-			task()
+			writer := task()
+
+			if writer != nil {
+				if isWriteLeader {
+					writer(ctx)
+				} else {
+					if first, hasFirst := writeQueue.LockedPeekFront(); hasFirst {
+						if time.Now().After(first.expiry) { // expired
+							writeQueue.LockedPopFront()
+						}
+					}
+					writeQueue.LockedPushBack(timedFunc{
+						expiry: time.Now().Add(retentionPeriod),
+						fn:     writer,
+					})
+				}
+			}
+		case newValue := <-setWriteLeader:
+			isWriteLeader = newValue
+			oldWriteQueue := writeQueue
+			writeQueue = channel.NewDeque[timedFunc](16)
+
+			for _, slice := range oldWriteQueue.LockedGetAll() {
+				for _, fn := range slice {
+					fn.fn(ctx)
+				}
+			}
 		}
 	}
 }
 
 func (ctrl *controller) Close(ctx context.Context) error { return nil }
 
+func (ctrl *controller) broadcastWriteLeader(b bool) {
+	for _, ch := range ctrl.setWriteLeaderChans {
+		go func(ch chan<- bool, b bool) {
+			ch <- b
+		}(ch, b)
+	}
+}
+
 func (ctrl *controller) resyncMonitorsLoop(ctx context.Context) {
 	logger := ctrl.Logger.WithField("submod", "resync")
 
 	defer shutdown.RecoverPanic(logger)
 
+	writeElector, err := multileader.NewElector(
+		"kelemetry-diff-cache-writer",
+		ctrl.Logger.WithField("submod", "write-leader-elector"),
+		ctrl.Clock,
+		&ctrl.options.writeElectorOptions,
+		ctrl.Clients.TargetCluster(),
+		ctrl.Metrics,
+	)
+	if err != nil {
+		logger.WithError(err).Error("cannot create write leader elector")
+		return
+	}
+
+	go writeElector.Run(ctx, func(ctx context.Context) {
+		count := ctrl.isWriteLeader.Add(1)
+		if count == 1 {
```

Comment:
is there any case count is not 1 ?

Reply:
It takes 15 seconds from context cancel (i.e. previous leader term ends) to `isWriteLeader` getting decremented. During this period, many things can happen, e.g. acquiring another leader term.